test/test_model.py

```diff
@@ -755,6 +759,243 @@ def assertBackendCalls(self, expected, *, reset=True):  # noqa: N802
         self.assertEqual(expected, self.harness._get_backend_calls(reset=reset))
 
 
+class PushPullCase:
+    """Test case for table-driven tests."""
+
+    def __init__(self, **vals):
+        self.pattern = None
+        self.dst = None
+        self.errorlist = []
+        self.wantlist = set()
+        for key in vals:
+            setattr(self, key, vals[key])
+
+
+recursive_list_cases = [
+    PushPullCase(
+        name='basic recursive list',
+        path='/',
+        files=['/foo/bar.txt', '/baz.txt'],
+        wantlist={'/foo/bar.txt', '/baz.txt'},
+    ),
+    PushPullCase(
+        name='basic recursive list reverse',
+        path='/',
+        files=['/baz.txt', '/foo/bar.txt'],
+        wantlist={'/baz.txt', '/foo/bar.txt'},
+    ),
+    PushPullCase(
+        name='directly list a (non-directory) file',
+        path='/baz.txt',
+        files=['/baz.txt'],
+        wantlist={'/baz.txt'},
+    ),
+]
+
+
+@pytest.mark.parametrize('case', recursive_list_cases)
+def test_recursive_list(case):
+    def list_func_gen(file_list):
+        args = {
+            'last_modified': datetime.time(),
+            'permissions': 0o777,
+            'size': 42,
+            'user_id': 0,
+            'user': 'foo',
+            'group_id': 1024,
+            'group': 'bar'}
```

Comment:
Style nit: it seems like we mostly add a trailing comma and put the end `}` on a separate line (see pebble.py, model.py, and tests). This is slightly better as the diff will only show a one-line change if you add an item to the end (and IMO more balanced looking, though that's definitely subjective). Same elsewhere with `FileInfo` constructors and the like.

Reply:
I also prefer the brace on a separate line.  Apparently autopep8 (which I used to do an autoformat) attaches it to the last entry.  fixed.